nucypher/blockchain/eth/agents.py

```diff
@@ -652,6 +652,17 @@ def fetch_policy(self, policy_id: str) -> list:
         blockchain_record = self.contract.functions.policies(policy_id).call()
         return blockchain_record
 
+    def fetch_policy_from_txid(self, txid, timeout=600):
```

Comment:
I think this function, instead of returning the policy creation event, should return the policy ID, and be renamed accordingly.

Although I made a comment somewhere below, where I say that instead of the txhash, Alice should directly send policy IDs, as it makes everyone' lives easier, so perhaps these functions are not relevant anymore.

Reply:
Also, I don't have an answer about @tuxxy's question, but regardless if it's a problem or not, policy IDs don't have this problem.

Reply:
@vzotova Could you answer my question?

Reply:
As far as I know all events accessible from any node because it's part of blocks
Also I agree with @cygnusv and want to use events only for monitoring and troubleshooting

Reply:
Is there another, better way to get a policy from a transaction?

Reply:
Like, does `self.blockchain.client.w3.eth.getTransaction(txid)` help me if, for my own hand-waving, I really want to use the tx?  Does `data` give me what I need there?

Reply:
Only theory: `data` should contain `policyId`, first 4 bytes it's just signature of function, next 16 bytes - what we need, `policyId`

Reply:
Is there a proper API for extracting that?  Rather than just splitting by hand?

Reply:
found it:
something like that
```
signature, parameters = self.contract.decode_function_input(testerchain.client.w3.eth.getTransaction(tx).data)
policy_id = parameters['_policyId']
```

Reply:
And in the same way, you can get the full list of nodes since they are also included in the call: `address[] calldata _nodes`.

Reply:
Is there also a way I can get the arrangements directly in the same call?

Reply:
I don't know exactly what you mean by "call" and by "arrangement". With respect to the latter, they mean different things in `nucypher` than in `PolicyManager`, where's mainly related to nodes' compensation. If with the former you mean "only calling the blockchain client once", I'd say that no, there's no way to get the arrangements.

Reply:
As you can see, we now solve this without using events, and without needing to get-and-reuse the policyID.  Hopefully all three of our concerns are addressed?